lib/Db/MessageMapper.php

```diff
@@ -373,24 +373,125 @@ public function updateBulk(Account $account, bool $permflagsEnabled, Message ...
+					$updateData["flag_notjunk_true"][] = $message->getUid();
+				} else {
+					$updateData["flag_notjunk_false"][] = $message->getUid();
+				}
+
+				if ($message->getFlagMdnsent()) {
+					$updateData["flag_mdnsent_true"][] = $message->getUid();
+				} else {
+					$updateData["flag_mdnsent_false"][] = $message->getUid();
+				}
+
+				if ($message->getFlagImportant()) {
+					$updateData["flag_important_true"][] = $message->getUid();
+				} else {
+					$updateData["flag_important_false"][] = $message->getUid();
+				}
+			}
+		}
+
+
 		try {
-			$query = $this->db->getQueryBuilder();
-			$query->update($this->getTableName())
-				->set('flag_answered', $query->createParameter('flag_answered'))
-				->set('flag_deleted', $query->createParameter('flag_deleted'))
-				->set('flag_draft', $query->createParameter('flag_draft'))
-				->set('flag_flagged', $query->createParameter('flag_flagged'))
-				->set('flag_seen', $query->createParameter('flag_seen'))
-				->set('flag_forwarded', $query->createParameter('flag_forwarded'))
-				->set('flag_junk', $query->createParameter('flag_junk'))
-				->set('flag_notjunk', $query->createParameter('flag_notjunk'))
-				->set('flag_mdnsent', $query->createParameter('flag_mdnsent'))
-				->set('flag_important', $query->createParameter('flag_important'))
-				->set('updated_at', $query->createNamedParameter($this->timeFactory->getTime()))
-				->where($query->expr()->andX(
-					$query->expr()->eq('uid', $query->createParameter('uid')),
-					$query->expr()->eq('mailbox_id', $query->createParameter('mailbox_id'))
-				));
+			// UPDATE messages SET flag true/false WHERE uid in (uids) -> for each flag
+			// => total of 20 queries
+			foreach ($flags as $flag) {
+				$queryTrue = $this->db->getQueryBuilder();
+				$queryTrue->update($this->getTableName())
+					->set($flag, $queryTrue->createNamedParameter(1, IQueryBuilder::PARAM_INT))
+					->set('updated_at', $queryTrue->createNamedParameter($this->timeFactory->getTime(), IQueryBuilder::PARAM_INT))
+					->where($queryTrue->expr()->andX(
+						$queryTrue->expr()->in('uid', $queryTrue->createParameter('uids')),
+						$queryTrue->expr()->eq('mailbox_id', $queryTrue->createNamedParameter($mailboxId, IQueryBuilder::PARAM_INT))
+					));
+				foreach (array_chunk($updateData[$flag.'_true'], 1000) as $chunk) {
+					$queryTrue->setParameter('uids', $chunk, IQueryBuilder::PARAM_INT_ARRAY);
+					$queryTrue->executeStatement();
+				}
+
+				$queryFalse = $this->db->getQueryBuilder();
+				$queryFalse->update($this->getTableName())
+					->set($flag, $queryFalse->createNamedParameter(0, IQueryBuilder::PARAM_INT))
+					->set('updated_at', $queryFalse->createNamedParameter($this->timeFactory->getTime(), IQueryBuilder::PARAM_INT))
+					->where($queryFalse->expr()->andX(
+						$queryFalse->expr()->in('uid', $queryFalse->createParameter('uids')),
+						$queryFalse->expr()->eq('mailbox_id', $queryFalse->createNamedParameter($mailboxId, IQueryBuilder::PARAM_INT))
```

Comment:
to experiment with: `AND flag=1`

to lower the number of accessed rows and hopefully the transaction size (for rollbacks and for replication)

Reply:
Seems worth pursuing 

![image](https://github.com/nextcloud/mail/assets/1374172/1c93703a-b26b-4e15-adac-32f1f27b5fb8)

Diff to the current HEAD

![image](https://github.com/nextcloud/mail/assets/1374172/1a543459-8e26-465e-9a9a-364e8ce2f0d5)

https://blackfire.io/profiles/compare/797f6164-4438-4c43-bcde-8d18dbd18b76/graph

Diff to main

![image](https://github.com/nextcloud/mail/assets/1374172/8494cd94-276a-4597-8adb-142e95454e06)

https://blackfire.io/profiles/compare/eaa6c793-1f5b-4793-b13d-ea14a5b9ca20/graph

Reply:
:white_check_mark: